app/models/product.rb

```diff
@@ -33,6 +34,16 @@ class Product
   validate :valid_measure_ids?
   validates :vendor, presence: true
 
+  def effective_date
+    # If selected, move effective date forward to the end of the actual reporting period
+    shift_records ? (Time.at(measure_period_start).in_time_zone + 1.year - 1.second).to_i : bundle.effective_date
+  end
+
+  def measure_period_start
```

Comment:
I'd suggest putting this method before `effective_date` for clarity into where the start_date_offset is coming in.